orchestra/static/orchestra/common/components/checklist/js/directives.js

```diff
@@ -0,0 +1,243 @@
+(function () {
+  'use strict';
+
+  angular
+    .module('orchestra.common.components.directives')
+    .directive('orchestraChecklist', ['$http', '$timeout', '$compile', 'orchestraService', orchestraChecklist]);
+
+  function orchestraChecklist($http, $timeout, $compile, orchestraService) {
+    return {
+      restrict: 'E',
+      scope: {
+        'data': '=',
+        'author': '=?',
+        'imagePrefix': '=?',
+      },
+      link: function(scope, el, attr) {
+        var sampleItem = {
+          'title': 'This is a sample checklist item. ' +
+                   'Double-click on the text to edit, ' +
+                   'or click to the right to expand.',
+          'comments': [
+                        {
+                          'author': '',
+                          'text': 'This is a sample checklist comment. ' +
+                                  'You can add a new comment for each new ' +
+                                  'round of iteration on your work!'
+                        }
+                      ],
+          'checked': false,
+          'expanded': false,
+          'readonly': true,
+          'index': 0,
+          'order': 0
+        };
+
+        // Initialize checklist data for a fresh task
+        scope.data = scope.data || {};
+        scope.data.items = scope.data.items || [sampleItem];
+        scope.data.deletedItems = scope.data.deletedItems || [];
+
+        orchestraService.signals.registerSignal('submit.before', function() {
+          scope.data.items.forEach(function(item) {
+            if (item.comments[item.comments.length - 1].text) {
+              item.comments[item.comments.length - 1].author = scope.author;
+              item.comments[item.comments.length - 1].timestamp = Date.now();
+              item.comments.push({'author': '', 'text': ''});
+            }
+          });
+        });
+
+        // Revert commit of latest comments if submit fails
+        orchestraService.signals.registerSignal('submit.error', function() {
+          scope.data.items.forEach(function(item) {
+            if (item.comments[item.comments.length - 1].text) {
+              item.comments.pop();
+            }
+          });
+        });
+
+        // Setup drag to reorder
+        scope.preventAction = false;
+        var checklistContainer = el.find('ul.checklist');
+        var drake = dragula([checklistContainer.get(0)], {
```

Comment:
It wasn't clear to me how the data is persisted in angular-dragula, but it seems like it just stores the HTML contents of the draggable containers. I definitely could be wrong, though. Also, you have to register it in the orchestra app definition rather than only in the steps that use it.